pkg/volume/emptydir/empty_dir.go

```diff
@@ -107,17 +110,58 @@ func (plugin *emptyDirPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, opts vo
 	return plugin.newMounterInternal(spec, pod, plugin.host.GetMounter(plugin.GetPluginName()), &realMountDetector{plugin.host.GetMounter(plugin.GetPluginName())}, opts)
 }
 
+func (plugin *emptyDirPlugin) calculateEmptyDirMemorySize(spec *volume.Spec, pod *v1.Pod) (*resource.Quantity, error) {
+	// if feature is disabled, continue the default behavior of linux host default
+	sizeLimit := &resource.Quantity{}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SizeMemoryBackedVolumes) {
+		return sizeLimit, nil
+	}
+
+	// size limit defaults to node allocatable (pods cant consume more memory than all pods)
+	nodeAllocatable, err := plugin.host.GetNodeAllocatable()
+	if err != nil {
+		return nil, err
+	}
+	sizeLimit = nodeAllocatable.Memory()
+
+	// determine pod resource allocation
+	// we use the same function for pod cgroup assigment to maintain consistent behavior
+	// NOTE: this could be nil on systems that do not support pod memory containment (i.e. windows)
+	podResourceConfig := cm.ResourceConfigForPod(pod, false, uint64(100000))
+	if podResourceConfig != nil && podResourceConfig.Memory != nil {
+		podMemoryLimit := resource.NewQuantity(*(podResourceConfig.Memory), resource.BinarySI)
+		// handle case where pod has no limit specified
+		if !podMemoryLimit.IsZero() && podMemoryLimit.Cmp(*sizeLimit) < 1 {
+			sizeLimit = podMemoryLimit
+		}
+	}
+
+	// volume local size is  used if and only if less than what pod could consume
+	if spec.Volume.EmptyDir.SizeLimit != nil {
+		volumeSizeLimit := spec.Volume.EmptyDir.SizeLimit
+		if volumeSizeLimit.Cmp(*sizeLimit) < 1 {
+			sizeLimit = volumeSizeLimit
+		}
+	}
+	return sizeLimit, nil
+}
+
 func (plugin *emptyDirPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, mounter mount.Interface, mountDetector mountDetector, opts volume.VolumeOptions) (volume.Mounter, error) {
 	medium := v1.StorageMediumDefault
-
+	sizeLimit := &resource.Quantity{}
 	if spec.Volume.EmptyDir != nil { // Support a non-specified source as EmptyDir.
 		medium = spec.Volume.EmptyDir.Medium
+		size, err := plugin.calculateEmptyDirMemorySize(spec, pod)
```

Comment:
We should only call this if the medium is memory, right?

Reply:
Ha, yeah.  Good catch :-)